flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -120,9 +121,8 @@ void setNext(StreamOperatorWrapper next) {
      * MailboxExecutor#yield()} to take the mails of closing operator and running timers and run
      * them.
      */
-    public void finish(StreamTaskActionExecutor actionExecutor, boolean isStoppingBySyncSavepoint)
-            throws Exception {
-        if (!isHead && !isStoppingBySyncSavepoint) {
+    public void finish(StreamTaskActionExecutor actionExecutor) throws Exception {
+        if (!isHead || wrapped instanceof StreamSource) {
```

Comment:
`StreamSource` getting `endInput()` call? Do we need that?

If yes, can we place this logic inside `SourceStreamTask` maybe? It will be easier to clean up the code once we drop legacy sources.